### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by kizaonline**

### CODE REVIEW

File: src/blob-uploader.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      if (!response.ok) {
+        return Promise.reject(new Error('Error calling blob api service: ' + response.status + ' ' + response.statusText))
+      }
+      return response.json()
+    })
+    .then((apiResponse) => {
+      this._uploadToS3(blob, apiResponse.putUrl)
+      return apiResponse.id
+    })
+    .catch((err) => Promise.reject(new Error('Error calling blob api service: ' + JSON.stringify(err))))
```
Comment (mymattcarroll) [inline src/blob-uploader.js]: `JSON.stringify(err)` will always return `{}`

See https://www.npmjs.com/package/serialize-error

I would instead just use `'' + err` as the `Error.prototype.toString()` returns `err.message`

File: src/blob-uploader.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    })
+  })
+
+  return fetch(request)
+    .then((response) => {
+      if (!response.ok) {
+        return Promise.reject(Error('Error uploading to S3: ' + response.status + ' ' + response.statusText))
+      }
+    })
+    .catch((err) => Promise.reject(new Error('Error uploading to S3: ' + JSON.stringify(err))))
```
Comment (mymattcarroll) [inline src/blob-uploader.js]: Same as above

File: src/blob-uploader.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  return fetch(request)
+    .then((response) => {
+      if (!response.ok) {
+        return Promise.reject(new Error('Error calling blob api service: ' + response.status + ' ' + response.statusText))
+      }
+      return response.json()
+    })
+    .then((apiResponse) => apiResponse.getUrl)
+    .catch((err) => Promise.reject(new Error('Error calling blob api service: ' + JSON.stringify(err))))
```
Comment (mymattcarroll) [inline src/blob-uploader.js]: same as above

File: src/blob-uploader.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+blobUploader.prototype.uploadBlob = function (
+  blob /*: Blob */
+) /* :Promise<number> */ {
+  if (!blob) {
+    return Promise.reject(new Error('blob argument not provided'))
+  }
+
+  if (!privateVars || !privateVars.get(this)) {
+    return Promise.reject(new Error('blobUploader uri not configured'))
+  }
```
Comment (mymattcarroll) [inline src/blob-uploader.js]: I think this check `!privateVars.get(this)` is being repeated on next `if` statement

Comment (kizaonline) [inline src/blob-uploader.js]: Correct and removed, think it came about trying to fix flow issues

File: src/blob-uploader.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  const request = new Request(vars.uri, {
+    method: 'POST',
+    mode: 'cors'
+  })
+
+  return fetch(request)
+    .then((response) => {
+      if (!response.ok) {
+        return Promise.reject(new Error('Error calling blob api service: ' + response.status + ' ' + response.statusText))
```
Comment (mymattcarroll) [inline src/blob-uploader.js]: This error will be caught in the catch below, meaning the error that will be returned to the caller of this function will be `Error calling blob api service: Error calling blob api service: 404 Not Found`

Instead i would reject here with `new Error(response.status + ' ' + response.statusText)`

Same applies to other places rejecting based on `response.ok`

**Comment by mymattcarroll** [comment]
LGTM 👍

**Pull request merged**